bgpd/bgp_io.c

```diff
@@ -52,8 +52,8 @@ static bool validate_header(struct peer *);
 
 /* Start and stop routines for I/O pthread + control variables
  * ------------------------------------------------------------------------ */
-_Atomic bool bgp_io_thread_run;
-_Atomic bool bgp_io_thread_started;
+static volatile bool bgp_io_thread_run = false;
+static volatile bool bgp_io_thread_started = false;
```

Comment:
I think `_Atomic` is better here as it clearly portrays the intent and usage semantics of these variables. Sure, in practice, on most (all?) platforms, a boolean will be a single byte, which "should" be updated in one go regardless of the use of explicit atomic operations, but relying on that is just bad form -- which is why I used `_Atomic` here to begin with. I don't see what this change is trying to rectify or improve; can you elaborate on your reasoning here?

Reply:
The current usage of atomic is overkill and was really implementing a cross thread mutex --  as we've seen, the implementation of a spin lock also caused issues.  I see this as more than a style issue.  A semaphore is the right os construct, but would be the sole place in FRR that uses it -- if you really object to the use of volatile, I'll submit the change to a semaphore.

Reply:
>The current usage of atomic is overkill

I disagree for the reasons outlined in my previous comment.

>and was really implementing a cross thread mutex -- as we've seen, the implementation of a spin lock also caused issues.

It caused one issue, which Donald fixed.

>A semaphore is the right os construct, but would be the sole place in FRR that uses it

Agreed on both counts.

>if you really object to the use of volatile, I'll submit the change to a semaphore.

Substituting volatile where you really mean atomic does not make sense to me, and is [incorrect](https://www.kernel.org/doc/html/latest/process/volatile-considered-harmful.html). From the same article:

>Another situation where one might be tempted to use volatile is when the processor is busy-waiting on the value of a variable. The right way to perform a busy wait is:
>```
>while (my_variable != what_i_want)
>    cpu_relax();
>```

Which is precisely what we have with Donald's fix in place.

In short -- what exactly is so egregious about using an atomic here?

Reply:
semaphore version now pushed.